pkg/p2p/internal/send_chan.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	cerror "github.com/pingcap/tiflow/pkg/errors"
+	proto "github.com/pingcap/tiflow/proto/p2p"
+	"go.uber.org/zap"
+)
+
+// SendChan is a specialized channel used to implement
+// the asynchronous interface of the MessageClient.
+// SendChan is a MPSC channel.
+type SendChan struct {
+	mu      sync.Mutex
+	buf     []*proto.MessageEntry
+	sendIdx int64
+	recvIdx int64
+
+	// hasNewMsg is a buffered channel that the sending goroutine(s)
+	// use to notify the receiver end.
+	hasNewMsg chan struct{}
+
+	// notifyChan is an unbuffered channel that
+	// SendSync writes to in order to determine whether the
+	// consumer is ready.
+	notifyChan chan struct{}
```

Comment:
Where do we write to this channel?

Reply:
It was a typo. Thank you for pointing this out. Fixing it has actually improved the performance of `SendSync` more than `1000` times.